data_structures/linked_list.cpp

```diff
@@ -1,134 +1,240 @@
+ */
+class list {
+ public:
+    list() : first(new link), last(first) {}
+
+    bool isEmpty();
+
+    void push_back(int new_elem);
+    void push_front(int new_elem);
+    void erase(int old_elem);
+    void display();
+    void search(int find_elem);
+    void reverse();
+
+ private:
+    std::shared_ptr<link> first;  ///< link before the actual first element
+    std::shared_ptr<link> last;   ///< last link on the list
+};
+
+/**
+ * function checks if list is empty
+ * @returns true if list is empty
+ * @returns false if list is not empty
+ */
+bool list::isEmpty() {
+    if (first == last) {
+        return true;
+    } else {
+        return false;
     }
+}
 
-    parent->next = temp->next;
-    delete temp;
+/**
+ * function adds new element to the end of the list
+ * @param new_elem to be added to the end of the list
+ */
+void list::push_back(int new_elem) {
+    std::shared_ptr<link> new_link(new link(new_elem));
+    last->succ = new_link;
+    last = last->succ;
 }
 
-void search(int x) {
-    node *t = start;
-    int found = 0;
-    while (t != NULL) {
-        if (t->val == x) {
-            std::cout << "\nFound";
-            found = 1;
-            break;
-        }
-        t = t->next;
+/**
+ * function adds new element to the beginning of the list
+ * @param new_elem to be added to front of the list
+ */
+void list::push_front(int new_elem) {
+    std::shared_ptr<link> new_link(new link(new_elem));
+    new_link->succ = first->succ;
+    first->succ = new_link;
+}
+
+/**
+ * function erases old element from the list
+ * @param old_elem to be erased from the list
+ */
+void list::erase(int old_elem) {
+    if (isEmpty()) {
+        std::cout << "List is Empty!";
+        return;
+    }
+    std::shared_ptr<link> t = first;
+    while (t != last && t->succ->val != old_elem) {
+        t = t->succ;
     }
-    if (found == 0) {
-        std::cout << "\nNot Found";
+    if (t == last) {
+        std::cout << "Element not found\n";
+        return;
+    }
+    t->succ = t->succ->succ;
+    if (t->succ == nullptr) {
+        last = t;
     }
 }
 
-void show() {
-    node *t = start;
-    while (t != NULL) {
-        std::cout << t->val << "\t";
-        t = t->next;
+/**
+ * function displays all the elements in the list
+ */
+void list::display() {
+    if (isEmpty()) {
+        std::cout << "List is Empty!";
+        return;
+    }
+    std::shared_ptr<link> t = first;
+    while (t->succ != nullptr) {
+        std::cout << t->succ->val << "\t";
+        t = t->succ;
     }
 }
 
-void reverse() {
-    node *first = start;
-    if (first != NULL) {
-        node *second = first->next;
-        while (second != NULL) {
-            node *tem = second->next;
-            second->next = first;
-            first = second;
-            second = tem;
-        }
-        start->next = NULL;
-        start = first;
-    } else {
-        std::cout << "\nEmpty list";
+/**
+ * function searchs for @param find_elem in the list
+ */
```

Comment:
1. missing `@param` description
2. after searching, should it return something back to the calling function? say pointer to the instance if found or `nullptr` if not found?